packages/firestore/src/local/shared_client_state.ts

```diff
@@ -0,0 +1,422 @@
+    this.clientStateKeyRe = new RegExp(
+      `^${CLIENT_STATE_KEY_PREFIX}_${persistenceKey}_([^_]*)$`
+    );
+  }
+
+  /** Returns 'true' if LocalStorage is available in the current environment. */
+  static isAvailable(): boolean {
+    return typeof window !== 'undefined' && window.localStorage != null;
+  }
+
+  start(knownClients: ClientKey[]): void {
+    assert(!this.started, 'WebStorageSharedClientState already started');
+    window.addEventListener('storage', this.storageListener);
+
+    for (const clientKey of knownClients) {
+      const storageKey = this.toLocalStorageClientKey(clientKey);
+      const clientState = RemoteClientState.fromLocalStorageEntry(
+        clientKey,
+        this.storage.getItem(storageKey)
+      );
+      if (clientState) {
+        this.activeClients[clientState.clientKey] = clientState;
+      }
+    }
+
+    this.started = true;
+    this.persistState();
+  }
+
+  getMinimumGlobalPendingMutation(): BatchId | null {
+    let minMutationBatch = null;
+    objUtils.forEach(this.activeClients, (key, value) => {
+      if (minMutationBatch === null) {
+        minMutationBatch = value.minMutationBatchId;
+      } else {
+        minMutationBatch = Math.min(value.minMutationBatchId, minMutationBatch);
+      }
+    });
+
+    return minMutationBatch;
+  }
+
+  getAllActiveQueryTargets(): SortedSet<TargetId> {
+    let activeTargets = new SortedSet<TargetId>(primitiveComparator);
+    objUtils.forEach(this.activeClients, (key, value) => {
+      activeTargets = activeTargets.unionWith(value.activeTargetIds);
+    });
+    return activeTargets;
+  }
+
+  addLocalPendingMutation(batchId: BatchId): void {
+    this.localClientState.addPendingMutation(batchId);
+    this.persistState();
+  }
+
+  removeLocalPendingMutation(batchId: BatchId): void {
+    this.localClientState.removePendingMutation(batchId);
+    this.persistState();
+  }
+
+  addLocalQueryTarget(targetId: TargetId): void {
+    this.localClientState.addQueryTarget(targetId);
+    this.persistState();
+  }
+
+  removeLocalQueryTarget(targetId: TargetId): void {
+    this.localClientState.removeQueryTarget(targetId);
+    this.persistState();
+  }
+
+  shutdown(): void {
+    assert(
+      this.started,
+      'WebStorageSharedClientState.shutdown() called when not started'
+    );
+    window.removeEventListener('storage', this.storageListener);
+    this.storage.removeItem(this.storageKey);
+    this.started = false;
+  }
+
+  private handleStorageEvent(event: StorageEvent): void {
+    if (!this.started) {
+      return;
+    }
+    if (event.storageArea === this.storage) {
+      // TODO(multitab): This assert will likely become invalid as we add garbage
+      // collection.
+      assert(
+        event.key !== this.storageKey,
+        'Received LocalStorage notification for local change.'
+      );
+      const clientKey = this.fromLocalStorageClientKey(event.key);
+      if (clientKey) {
+        if (event.newValue == null) {
```

Comment:
Consider adding logging for these LocalStorage notifications (can do in a later PR, e.g. when adding callbacks to SyncEngine if you prefer).

Reply:
I think we would want to log the actual notifications that we deliver to SyncEngine, so I am going to defer on this until we get to that part.